text/2018-11-12-backup-restore-replication-with-raft-learner.md

```diff
@@ -0,0 +1,168 @@
+
+```
++---------+                                      
+|Node A   |                                      
++---------+----+                          +-----+
+|Region 1*|    |                     +----|Ceph |
++---------+    |      +---------+    |    +-----+
+               |      |Backup   |    |           
++---------+    |      +---------+    |           
+|Node B   |    |      |Region 1 |    |    +-----+
++---------+-----------+---------+---------|S3   |
+|Region 2*|    |      |Region 2 |    |    +-----+
++---------+    |      +---------+    |           
+               |      |Region 3 |    |           
++---------+    |      +---------+    |    +-----+
+|Node C   |    |                     +--- |Disk |
++---------+----+                          +-----+
+|Region 3*|                                      
++---------+                                                                                   
+ ```
+
+There may be a problem when the Backup node is first added to the cluster. Many snapshots will be created and this will also cause a high load for the Raft leaders. We can alleviate this problem as below:
+
+1. Let the PD limit the speed. 
+2. Introduce the [Follower Snapshot](https://github.com/pingcap/raft-rs/issues/135) mechanism. When a new learner is added, the leader won’t send a snapshot to the learner. Instead, the leader will ask a follower to generate the snapshot and send the snapshot to the new learner. After the new learner receives the snapshot and applies it, the leader will replicate the logs to the new learner again.
+3. Furthermore, we can even introduce [Follower replication](https://github.com/pingcap/raft-rs/issues/136) to let the follower do the log replication directly - to reduce the leader pressure. 
+
+Unlike common Raft peers, the learner in the Backup node doesn't apply the Raft logs, it only saves some Region information in the local RocksDB. The snapshot files and Raft logs will be saved to the distributed file system like Ceph or object storage like S3,
+
+**In the following example, we will still use Leader for snapshot and replication, and use Ceph to represent the remote storage.**
+
+When a new learner of Region 1 is added to the Backup node, it will:
+
+1. Create a Raft peer and save the Region meta in the local RocksDB.
+2. Receive the snapshot from the leader and save it to the Ceph.
+3. Persist the Raft state (like applied index, committed Index) in the local RocksDB. Of course, we must also record how to find the snapshot file in Ceph :-)
+4. Receive the Raft logs from the leader, and append them to the Ceph. For simplicity, we can use one file for one Region and append the logs to it. To ensure higher safety and to avoid our foolish mistakes like writing wrong data to the state machine, we need to keep Raft logs as more as possible, maybe keep all logs and don’t compact them. 
```

Comment:
```suggestion
4. Receive the Raft logs from the leader, and append them to the Ceph. For simplicity, we can use one file for one Region and append the logs to it. To ensure higher safety and to avoid mistakes like writing wrong data to the state machine, we need to keep Raft logs as many as possible, maybe keep all logs and don’t compact them. 
```

Reply:
many -> much :)